design/wasi_unstable/wasi_unstable.witx

```diff
@@ -0,0 +1,525 @@
+    (param $fd $fd_t)
+    (result $error $errno_t)
+  )
+
+  ;; Get the attributes of a file descriptor.
+  ;; Note: This returns similar flags to `fsync(fd, F_GETFL)` in POSIX, as well as additional fields.
+  (@interface func (export "fd_fdstat_get")
+    (param $fd $fd_t)
+    (result $error $errno_t)
+    ;; The buffer where the file descriptor's attributes are stored.
+    (result $fdstat_t)
+  )
+
+  ;; Adjust the flags associated with a file descriptor.
+  ;; Note: This is similar to `fcntl(fd, F_SETFL, flags)` in POSIX.
+  (@interface func (export "fd_fdstat_set_flags")
+    (param $fd $fd_t)
+    ;; The desired values of the file descriptor flags.
+    (param $flags $fdflags)
+    (result $error $errno_t)
+  )
+
+  ;;  Adjust the rights associated with a file descriptor.
+  ;; This can only be used to remove rights, and returns `ENOTCAPABLE` if called in a way that would attempt to add rights
+  (@interface func (export "fd_fdstat_set_rights")
+    (param $fd $fd_t)
+    ;; The desired rights of the file descriptor.
+    (param $fs_rights_base $rights_t)
+    (param $fs_rights_inheriting $rights_t)
+    (result $error $errno_t)
+  )
+
+  ;; Return the attributes of an open file.
+  (@interface func (export "fd_filestat_get")
+    (param $fd $fd_t)
+    (result $error $errno_t)
+    ;; The buffer where the file's attributes are stored.
+    (result $fs_rights_base $rights_t)
+  )
+
+  ;; Adjust the size of an open file. If this increases the file's size, the extra bytes are filled with zeros.
+  ;; Note: This is similar to `ftruncate` in POSIX.
+  (@interface func (export "fd_filestat_set_size")
+    (param $fd $fd_t)
+    ;; The desired file size.
+    (param $st_size $filesize_t)
+    (result $error $errno_t)
+  )
+
+  ;; Adjust the timestamps of an open file or directory.
+  ;; Note: This is similar to `futimens` in POSIX.
+  (@interface func (export "fd_filestat_set_times")
+    (param $fd $fd_t)
+    ;; The desired values of the data access timestamp.
+    (param $st_atim $timestamp_t)
+    ;; The desired values of the data modification timestamp.
+    (param $st_mtim $timestamp_t)
+    ;; A bitmask indicating which timestamps to adjust.
+    (param $fst_flags $fstflags_t)
+    (result $error $errno_t)
+  )
+
+  ;; Read from a file descriptor, without using and updating the file descriptor's offset.
+  ;; Note: This is similar to `preadv` in POSIX.
+  (@interface func (export "fd_pread")
+    (param $fd $fd_t)
+    ;; List of scatter/gather vectors in which to store data.
+    (param $iovs $iovec_t_array)
+    ;; The offset within the file at which to read.
+    (param $offset $filesize_t)
+    (result $error $errno_t)
+    ;; The number of bytes read.
+    (result $nread $size_t)
+  )
+
+  ;; Return a description of the given preopened file descriptor.
+  (@interface func (export "fd_prestat_get")
+    (param $fd $fd_t)
+    (result $error $errno_t)
+    ;; The buffer where the description is stored.
+    (result $buf $prestat_t)
```

Comment:
It looks like `prestat_t` is not defined (related to https://github.com/WebAssembly/WASI/issues/51).

Reply:
Good spot! I've now added it.